sqlglot/dialects/bigquery.py

```diff
@@ -184,6 +185,16 @@ def _array_contains_sql(self: BigQuery.Generator, expression: exp.ArrayContains)
     )
 
 
+def _ts_or_ds_add_sql(self: BigQuery.Generator, expression: exp.TsOrDsAdd) -> str:
+    return date_add_interval_sql("DATE", "ADD")(self, ts_or_ds_add_cast(expression))
+
+
+def _ts_or_ds_diff_sql(self: BigQuery.Generator, expression: exp.TsOrDsDiff) -> str:
+    expression.this.replace(exp.cast(expression.this, "TIMESTAMP", copy=True))
+    expression.expression.replace(exp.cast(expression.this, "TIMESTAMP", copy=True))
```

Comment:
The reasoning here is that `DATEDIFF` [returns](https://cloud.google.com/bigquery/docs/reference/standard-sql/date_functions#date_diff) a `BIGINT` in both BigQuery and Redshift, so promoting the args `TIMESTAMP` shouldn't really affect the result, unless the value to-be-cast is a _time_ string because then the cast will fail. But then again, we can't really handle all cases without annotating types.